### PR TITLE
✨ Command LDAP

### PR DESCRIPTION
**Pull request by Nathan-Al**
> Please provide enough information so that others can review your pull request:

:sparkles: Manage LDAP server through Command lines

> Explain the **details** for making this change. What existing problem does the pull request solve?

> Screenshots/GIFs

### CODE REVIEW

File: app/src/Command/LdapCreateEntry.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $dn = $input->getArgument('dn');
+        $attributes = $input->getArgument('attr');
+
+        $symfonyStyle = new SymfonyStyle($input, $output);
+
+        $jsonDecodeAttributes = json_decode($attributes, true);
+
+        if ($jsonDecodeAttributes==null) {
```
Comment (madmath03) [inline app/src/Command/LdapCreateEntry.php:65]: You need to check that the result is an array and not empty, not just null

File: app/src/Command/LdapCreateEntry.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            $symfonyStyle->error("The Attribute argument is not a valid JSON.");
```
Comment (madmath03) [inline app/src/Command/LdapCreateEntry.php:66]: Do not use double quotes when simple quotes are enough.
```suggestion
            $symfonyStyle->error('The Attribute argument is not a valid JSON.');
```

File: app/src/Command/LdapCreateEntry.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            return 1;
+        }
+
+        if ($this->client->create($dn, $jsonDecodeAttributes)) {
+            $symfonyStyle->success('Following LDAP entry was successfuly create');
```
Comment (madmath03) [inline app/src/Command/LdapCreateEntry.php:71]: You mention a "_following_" entry but do not mention the DN. Also typo: missing `l` in "_successfuly_".

```suggestion
            $symfonyStyle->success("Following LDAP entry was successfully created: $dn");
```

File: app/src/Command/LdapDeleteEntry.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+    /**
+     * Configures the current command.
+     *
+     * @return void
+     */
+    protected function configure()
+    {
+        $this
+            ->setDescription('Delete a ldap Entry')
```
Comment (madmath03) [inline app/src/Command/LdapDeleteEntry.php:36]:
```suggestion
            ->setDescription('Delete a LDAP Entry')
```

File: app/src/Command/LdapDeleteEntry.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        $dn = $input->getArgument('dn');
+        $symfonyStyle = new SymfonyStyle($input, $output);
+
+        if ($this->client->delete($dn)) {
+            $symfonyStyle->success('Following LDAP entry was successfuly create');
+            return 0;
+        }
+        
+        $symfonyStyle->error("An error occurred during creation of LDAP entry");
```
Comment (madmath03) [inline app/src/Command/LdapDeleteEntry.php:55]: Do not use double quotes when simple quotes are enough. Also bad message copy pasted from create command.
```suggestion
        $symfonyStyle->error('An error occurred during deletion of LDAP entry');
```

File: app/src/Command/LdapUpdateEntry.php [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        $symfonyStyle->comment("update entry :");
+        
+        $jsonDecodeAttributes = json_decode($attributes, true);
+        
+        if ($jsonDecodeAttributes===null) {
+            $symfonyStyle->error('The attribute Option is not a valid JSON');
+            return 1;
+        }
+
+        $e = $this->client->update($dn, $jsonDecodeAttributes);
```
Comment (madmath03) [inline app/src/Command/LdapUpdateEntry.php:73]: Rename `$e` to `$entry`.
```suggestion
        $e = $this->client->update($dn, $jsonDecodeAttributes);
```

File: app/src/Command/LdapUpdateEntry.php [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        if ($e) {
+            $symfonyStyle->success('Following LDAP entry was successfuly updated');
```
Comment (madmath03) [inline app/src/Command/LdapUpdateEntry.php:75]: You mention a "_following_" entry but do not mention the DN. Also typo: missing `l` in "_successfuly_".

```suggestion
            $symfonyStyle->success("Following LDAP entry was successfully updated: $dn");
```

File: app/src/Service/Ldap/Client.php [hunk 8]
```diff
@@ -77,4 +79,77 @@ public function check(
...
+    public function update(string $query, array $attributes) : bool
+    {
+        $entryManager = $this->ldap->getEntryManager();
+
+        // Finding and updating an existing entry
+        $result = $this->executeQuery($query);
+
+        // FIXME Check result before doing anything on it
+        $entry = $result[0];
+        if ($entry===null) {
```
Comment (madmath03) [inline app/src/Service/Ldap/Client.php:117]: Not a valid test. If result is an empty array, `$entry` will be `undefined` not `null`. Use `empty` method instead.
```suggestion
        if (empty($entry)) {
```

File: composer.json [hunk 9]
```diff
@@ -0,0 +1,2 @@
+{
+}
```
Comment (madmath03) [inline composer.json:2]: What is this file ???

File: app/src/Service/Ldap/Client.php [hunk 10]
```diff
@@ -77,4 +79,77 @@ public function check(
...
+
+        $entry = new Entry($dn, $attributes);
+        $entryManager = $this->ldap->getEntryManager();
+
+        return $entryManager->add($entry);
+    }
+
+    /**
+     * @return bool|null
+     */
```
Comment (madmath03) [inline app/src/Service/Ldap/Client.php:107]: Make sure to document thrown exception from Symfony LDAP client.

File: app/src/Service/Ldap/Client.php [hunk 11]
```diff
@@ -77,4 +79,77 @@ public function check(
...
+            $entry->setAttribute($key, $value);
+        }
+        $entryManager->update($entry);
+
+        return true;
+    }
+
+    /**
+     * @return boolean
+     */
```
Comment (madmath03) [inline app/src/Service/Ldap/Client.php:131]: Make sure to document thrown exception from Symfony LDAP client.

File: app/src/Service/Ldap/Client.php [hunk 12]
```diff
@@ -77,4 +79,77 @@ public function check(
...
+     *
+     * @psalm-return \Symfony\Component\Ldap\Adapter\CollectionInterface|array<array-key, Entry>
+     */
+    public function executeQuery(string $query)
+    {
+        $this->ldap->bind($this->config['search_dn'], $this->config['search_password']);
+        return $this->ldap->query($this->config['base_dn'], $query)->execute();
+    }
+
+    public function create(string $dn, array $attributes): bool
```
Comment (madmath03) [inline app/src/Service/Ldap/Client.php:94]: Make sure to document thrown exception from Symfony LDAP client.

File: app/src/Command/LdapSearchEntry.php [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        $entries = $this->client->search($query);
+
+        $rows = [];
+        //Always put DN in labels
+        $labels = ['DN'];
+
+        // TODO Build array of attributes labels
+        foreach ($arrayAttributes as $attribute => $value) {
+            $labels[] = $attribute;
+        }
```
Comment (madmath03) [inline app/src/Command/LdapSearchEntry.php:90]: Just merge the arrays.
```suggestion
        $labels= array_merge($labels, $arrayAttributes);
```

File: app/src/Command/LdapSearchEntry.php [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+            if (is_array($entryAttrArray)) {
+                // Always put DN in rows
+                $rows[$entryDn] = [
+                    json_encode($entryDn),
+                ];
+                // Get the ldap entry by the number of attribute given in command
+                foreach ($arrayAttributes as $attribute => $value) {
+                    //TODO Verify if the entry is not empty if empty return null
+                    //TODO Fix the way is handle the verification
+                    if (!empty($entryAttrArray[$value])) {
```
Comment (madmath03) [inline app/src/Command/LdapSearchEntry.php:105]: Completely wrong test !
1. You need to check that `$entryAttrArray` is not empty
2. You need to check that `$entryAttrArray` contains the requested attribute

**Review by madmath03 — CHANGES_REQUESTED** [review]

File: composer.json (hunk 9 above)
Comment (Nathan-Al) [reply composer.json:2]: Auto generate, but i don't know by what. I delete him for the next commit.

File: app/src/Command/LdapCreateEntry.php (hunk 1 above)
Comment (Nathan-Al) [reply app/src/Command/LdapCreateEntry.php:65]: Done

File: app/src/Command/LdapCreateEntry.php (hunk 2 above)
Comment (Nathan-Al) [reply app/src/Command/LdapCreateEntry.php:66]: Done

File: app/src/Command/LdapCreateEntry.php (hunk 3 above)
Comment (Nathan-Al) [reply app/src/Command/LdapCreateEntry.php:71]: Done

File: app/src/Command/LdapCreateEntry.php [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+            return 0;
+        }
+        
+        $symfonyStyle->error("An error occurred during creation of LDAP entry");
```
[the remark this answers is not in the document]
Comment (Nathan-Al) [reply app/src/Command/LdapCreateEntry.php:75]: Done

File: app/src/Command/LdapUpdateEntry.php (hunk 6 above)
Comment (Nathan-Al) [reply app/src/Command/LdapUpdateEntry.php]: That line was not meant to be here, she has been deleted.

File: app/src/Command/LdapSearchEntry.php (hunk 13 above)
Comment (Nathan-Al) [reply app/src/Command/LdapSearchEntry.php]: Done

File: app/src/Command/LdapCreateEntry.php [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+        return 1;
+    }
+
+    protected function isValid(SymfonyStyle $ioStyle, $dn): bool
```
[the remark this answers is not in the document]
Comment (Nathan-Al) [reply app/src/Command/LdapCreateEntry.php]: Yes, i forgot to delete it, it's done now.

File: app/src/Command/LdapCreateEntry.php (hunk 1 above)
Comment (Nathan-Al) [reply app/src/Command/LdapCreateEntry.php]: j'ai laissé la conditions null car la fonction peut aussi renvoyer null.

**Review by madmath03 — CHANGES_REQUESTED** [review]

File: app/src/Command/LdapConfig.php [hunk 17]
```diff
@@ -0,0 +1,31 @@
+<?php
+
+namespace App\Command;
+
+trait LdapConfig
+{
+    public function returnConfig()
```
[the remark this answers is not in the document]
Comment (Nathan-Al) [reply app/src/Command/LdapConfig.php:7]: Ok

File: app/src/Command/LdapSearchEntry.php [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+
+        array_unshift($labels, 'dn');
+
+        foreach ($ldapEntries as $key => $entry) {
+            $entryDn = $entry->getDn();
+            $entries[$key]['dn'] = $entryDn;
+
+            foreach ($attributes as $attr) {
+                if ($entry->hasAttribute($attr) && !empty($entry->hasAttribute($attr))) {
+                    //Possibility to change json_encode by implode(', ', $var);
```
[the remark this answers is not in the document]
Comment (Nathan-Al) [reply app/src/Command/LdapSearchEntry.php:96]: I see, thank for the informations

File: app/src/Command/LdapCreateEntry.php [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+            'uid_key' => $uidKey,
+            'mail_key' => $mailKey,
+            'base_dn' => $baseDn,
+            'is_ad' => "0",
+            'ad_domain' => 'planetexpress.com',
+            'query' => $queryLdap,
+            'search_dn' => 'cn=admin,dc=planetexpress,dc=com',
+            'search_password' => 'GoodNewsEveryone',
+            'enabled' => '1'
+        ];
```
[the remark this answers is not in the document]
Comment (Nathan-Al) [reply app/src/Command/LdapCreateEntry.php:92]: I was testing the config, but i move it after.

File: app/src/Command/LdapCreateEntry.php [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+    {
+
+        $distingName = $input->getArgument('distingName');
+        $attributes = $input->getArgument('attr');
+
+        $symfonyStyle = new SymfonyStyle($input, $output);
+
+        $jsonDecodeAttributes = json_decode($attributes, true);
+        
+        if (is_array($jsonDecodeAttributes) && empty($jsonDecodeAttributes) || $jsonDecodeAttributes==null) {
```
[the remark this answers is not in the document]
Comment (Nathan-Al) [reply app/src/Command/LdapCreateEntry.php:70]: I see, ok

**Review by madmath03 — CHANGES_REQUESTED** [review]
Some improvements on the feature and description itself. Still need to review the Unit Tests though.

**Review by madmath03 — APPROVED** [review]

**Pull request merged**